test/test_tensorexpr.py

```diff
@@ -1560,5 +1560,121 @@ def foo(a, b, c):
             exp = traced(a, b, c)
             self.assertEqual(ref, exp)
 
+    def test_propagated_mem_layout(self):
+        def foo(a, b, c):
+            t_next = c + 1
+            t5 = t_next * b
+            t7 = a * t5
+            return t7
+
+        def foo_multi_outputs(a, b, c):
+            t_next = c + 1
+            t5 = b * t_next
+            t7 = a * t5
+            return (t7, t5, t_next)
+
+        def foo_multi_outputs_i_nhwc_o_nchw(a, b, c):
+            t_next = c + 1
+            t5 = b * t_next
+            t7 = a * t5
+            t8 = t7.to(memory_format=torch.contiguous_format)
+            return (t8, t7, t5, t_next)
+
+        def run_foo_case(foo, a, b, c):
+            traced_contiguous = torch.jit.trace(foo, (a, b, c))
+            ref = foo(a, b, c)
+            exp = traced_contiguous(a, b, c)
+            exp = traced_contiguous(a, b, c)
+            self.assertEqual(ref, exp)
+
+        for strategy in ["STATIC", "DYNAMIC"]:
+            old_strategy = torch.jit.set_fusion_strategy([(strategy, 10)])
+
+            # The memory layouts of the inputs are
+            #     (contiguous, contiguous, contiguous)
+            a = torch.rand(2, 3, 4, 5, dtype=torch.float32)
+            b = torch.rand(2, 3, 4, 5, dtype=torch.float32)
+            c = torch.rand(2, 3, 4, 5, dtype=torch.float32)
+            # The memory layout of the imperative mode output is
+            #     (contiguous)
+            # propagation memory layout is contiguous
+            run_foo_case(foo, a, b, c)
+            # The memory layouts of the imperative mode outputs are
+            #     (contiguous, contiguous, contiguous)
+            # propagation memory layout is contiguous
+            run_foo_case(foo_multi_outputs, a, b, c)
+
+            # The memory layouts of the inputs are
+            #     (contiguous, channels-last, contiguous)
+            a = torch.rand(2, 3, 4, 5, dtype=torch.float32)
+            b = torch.rand(2, 3, 4, 5, dtype=torch.float32).to(memory_format=torch.channels_last)
+            c = torch.rand(2, 3, 4, 5, dtype=torch.float32)
+            # The memory layout of the imperative mode output is
+            #     (contiguous)
+            # propagation memory layout is contiguous
+            run_foo_case(foo, a, b, c)
+            # The memory layouts of the imperative mode outputs are
+            #     (contiguous, channels-last, contiguous)
+            # propagation memory layout is contiguous
+            run_foo_case(foo_multi_outputs, a, b, c)
```

Comment:
This looks like a lot of duplication. Would it be possible to factor out the common code? E.g. we could have a "config" array and then iterate over it:
```
configs = [
  ([torch.contiguous, torch.channels_last, torch.contiguous], foo), 
  ([torch.contiguous, torch.channels_last, torch.contiguous], foo_multi_outputs), 
  ([torch.contiguous, torch.channels_last, torch.contiguous], foo_multi_outputs_i_nhwc_o_nchw), 
  ...
]
```